[SPARK-11787] [SQL] Improve Parquet scan performance when using flat schemas.

nongli opened the pull request:
This patch adds an alternate to the Parquet RecordReader from the parquet-mr project
that is much faster for flat schemas. Instead of using the general converter mechanism
from parquet-mr, this directly uses the lower level APIs from parquet-columnar and a
customer RecordReader that directly assembles into UnsafeRows.

This is optionally disabled and only used for supported schemas.

Using the tpcds store sales table and doing a sum of increasingly more columns, the results
are:

For 1 Column:
  Before: 11.3M rows/second
  After: 18.2M rows/second

For 2 Columns:
  Before: 7.2M rows/second
  After: 11.2M rows/second

For 5 Columns:
  Before: 2.9M rows/second
  After: 4.5M rows/second

nongli committed:
[SPARK-11787] [SQL] Improve Parquet scan performance when using flat schemas.

This patch adds an alternate to the Parquet RecordReader from the parquet-mr project
that is much faster for flat schemas. Instead of using the general converter mechanism
from parquet-mr, this directly uses the lower level APIs from parquet-columnar and a
customer RecordReader that directly assembles into UnsafeRows.

This is optionally disabled and only used for supported schemas.

Using a benchmark that just sums columns reading from a 1SF TPCDS store_sales table,
I measured the rates of the entire query:

For 1 Column:
  Before: 11.3M rows/second
  After: 18.2M rows/second

For 2 Columns:
  Before: 7.2M rows/second
  After: 11.2M rows/second

For 5 Columns:
  Before: 2.9M rows/second
  After: 4.5M rows/second

SparkQA commented:
**[Test build #46105 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46105/consoleFull)** for PR 9774 at commit [`56f86c5`](https://github.com/apache/spark/commit/56f86c5e7ef447738c5bfd67d1a58a8c1d755f0a).

SparkQA commented:
**[Test build #46105 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46105/consoleFull)** for PR 9774 at commit [`56f86c5`](https://github.com/apache/spark/commit/56f86c5e7ef447738c5bfd67d1a58a8c1d755f0a).
 * This patch **fails RAT tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `public abstract class SpecificParquetRecordReaderBase<T> extends RecordReader<Void, T> `\n  * `  abstract static class IntIterator `\n  * `  protected static final class ValuesReaderIntIterator extends IntIterator `\n  * `  protected static final class RLEIntIterator extends IntIterator `\n  * `  protected static final class NullIntIterator extends IntIterator `\n  * `public class UnsafeRowParquetRecordReader extends SpecificParquetRecordReaderBase<UnsafeRow> `\n

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46105/
Test FAILed.

rxin commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/UnsafeRowParquetRecordReader.java:
```diff
@@ -0,0 +1,428 @@
+ *
+ * TODO: handle complex types.
+ */
+public class UnsafeRowParquetRecordReader extends SpecificParquetRecordReaderBase<UnsafeRow> {
```
you might want to explain some pieces of this is taken from parquet-mr.

rxin commented on core/src/main/scala/org/apache/spark/rdd/SqlNewHadoopRDD.scala:
```diff
@@ -95,6 +95,8 @@ private[spark] class SqlNewHadoopRDD[V: ClassTag](
 
   @transient protected val jobId = new JobID(jobTrackerId, id)
+  protected val enableUnsafeRowParquetReader: Boolean =
+      sc.conf.getBoolean("spark.parquet.enableUnsafeRowRecordReader", true)
```
we should use a sqlconf to set this, not the spark conf (spark conf is immutable once the app starts)

nongli replied:
can i use that here? This seems far away from sql.

rxin replied:
We can move this file to sql.execution package and put SQLConf in here.

nongli replied:
Ok let me do that next. This is used in a few places and not very related to this patch.

julienledem commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/SpecificParquetRecordReaderBase.java:
```diff
@@ -0,0 +1,233 @@
+ * This class handles computing row groups, filtering on them, setting up the column readers,
+ * etc.
+ * TODO: move this to the parquet-mr project. There are performance benefits of doing it
+ * this way, albeit at a higher cost to implement.
```
This makes sense to me. We should have this in https://github.com/apache/parquet-mr/

julienledem commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/UnsafeRowParquetRecordReader.java:
```diff
@@ -0,0 +1,428 @@
+
+  private void decodeLongBatch(int col, int num) throws IOException {
+    for (int n = 0; n < num; ++n) {
+      if (columnReaders[col].next() != 0) {
```
how about computing all the indices first to avoid the branch?
```
for (int n = 0; n < num; ++n) {
  index += (1 - columnReaders[col].next())
  rowIndex[n] = index;
}
for (int n = 0; n < num; ++n) {
  rows[rowIndex[n]] = columnReaders[col].nextLong()
}
```

julienledem replied:
you can have the JIT print out the generated code.

julienledem commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/UnsafeRowParquetRecordReader.java:
```diff
@@ -0,0 +1,428 @@
+          break;
+        case DOUBLE:
+          decodeDoubleBatch(i, num);
+          break;
```
is this assuming no nulls?

nongli replied:
It handles NULLs. The batch decode reads the def levels.

julienledem commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/UnsafeRowParquetRecordReader.java:
```diff
@@ -0,0 +1,428 @@
+    private int next() throws IOException {
+      if (valuesRead >= endOfPageValueCount) {
+        if (valuesRead >= totalValueCount) {
+          // How do we get here? Throw end of stream exception?
```
we always read one more repetitionLevel (0) because the assembly algorithm depends on the next Repetition Level

nongli replied:
I see

julienledem commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/UnsafeRowParquetRecordReader.java:
```diff
@@ -0,0 +1,428 @@
+
+    private void readPage() throws IOException {
+      DataPage page = pageReader.readPage();
+      // TODO: Why is this a visitor?
```
backward compatibility for page formats

nongli replied:
Reading two page versions is fine but do we really need a visitor here?

julienledem replied:
well, you can use instanceof and casts, is that what you are getting at?
the visitor is type safe ;)
In Scala you would just use patter matching:
```
page match {
  case dataPageV1: DataPageV1 => readPageV1(dataPageV1)
  case dataPageV2: DataPageV2 => readPageV2(dataPageV2)
}
```

nongli replied:
It's type safe but it doesnt throw IOException which is annoying.

julienledem replied:
That's fair. We could add a type param for that (Visitor<R, T>... Void visit(page) throws T). But really, you'd want to reuse the existing code I think.
Your methods readPageV[1|2] already catch IOException and throw another one, so you could throw something else there too.
You could use a switch statement or instanceof but the visitor is the only way I know of that will make compilation fail if we add a 3rd page format.
suggestions welcome in any case.

julienledem commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/UnsafeRowParquetRecordReader.java:
```diff
@@ -0,0 +1,428 @@
+      } catch (IOException e) {
+        throw new IOException("could not read page " + page + " in col " + descriptor, e);
+      }
+    }
```
this should not need to be duplicated

nongli replied:
Agreed. When we think about moving this into parquet-mr, we can refactor that side a bit more.

julienledem replied:
feel free to refactor in parquet to make bits reusable.

SparkQA commented:
**[Test build #46196 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46196/consoleFull)** for PR 9774 at commit [`b6a3b62`](https://github.com/apache/spark/commit/b6a3b62bb32b414c5243e640c69a5482da78bfd6).

rxin commented:
Did you open two pull requests?

SparkQA commented:
**[Test build #46196 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46196/consoleFull)** for PR 9774 at commit [`b6a3b62`](https://github.com/apache/spark/commit/b6a3b62bb32b414c5243e640c69a5482da78bfd6).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `public abstract class SpecificParquetRecordReaderBase<T> extends RecordReader<Void, T> `\n  * `  abstract static class IntIterator `\n  * `  protected static final class ValuesReaderIntIterator extends IntIterator `\n  * `  protected static final class RLEIntIterator extends IntIterator `\n  * `  protected static final class NullIntIterator extends IntIterator `\n  * `public class UnsafeRowParquetRecordReader extends SpecificParquetRecordReaderBase<UnsafeRow> `\n

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46196/
Test FAILed.

marmbrus commented on core/src/main/scala/org/apache/spark/rdd/SqlNewHadoopRDD.scala:
```diff
@@ -96,6 +96,11 @@ private[spark] class SqlNewHadoopRDD[V: ClassTag](
+  // If true, enable using the custom RecordReader for parquet. This only works for
+  // a subset of the types (no complex types).
+  protected val enableUnsafeRowParquetReader: Boolean =
+      sc.conf.getBoolean("spark.parquet.enableUnsafeRowRecordReader", true)
```
Sorry, I said this on the wrong PR, but we should get this from the SQLConf since the Spark conf is immutable.

marmbrus commented on core/src/main/scala/org/apache/spark/rdd/SqlNewHadoopRDD.scala:
```diff
@@ -150,9 +155,30 @@ private[spark] class SqlNewHadoopRDD[V: ClassTag](
+      if (enableUnsafeRowParquetReader &&
+          format.getClass.getName == "org.apache.parquet.hadoop.ParquetInputFormat") {
+        reader = Utils.classForName(
+          "org.apache.spark.sql.execution.datasources.parquet.UnsafeRowParquetRecordReader")
```
Why are we using reflection to create an instance?

nongli replied:
This thing is in the wrong package. I'll clean this up when I move this class.

marmbrus commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/UnsafeRowParquetRecordReader.java:
```diff
@@ -0,0 +1,613 @@
+
+      originalTypes[i] = t.getOriginalType();
+
+      // TODO: Be extremely cautious in what is supported. Expand this.
```
Could we move this logic to where we are deciding which reader to instantiate instead of using exceptions for control flow?

nongli committed:
Use UnsafeRowWriter/BufferHolder for string data.

This path is going to be heavy to begin with so some of the other optimizations don't make sense.
Issues were exposed with the UTF-8 test cases with the prior approach.

This also adds more checks for supported schemas.

davies commented on core/src/main/scala/org/apache/spark/rdd/SqlNewHadoopRDD.scala:
```diff
@@ -30,10 +28,12 @@ import org.apache.spark.broadcast.Broadcast
-import org.apache.spark.storage.StorageLevel
-import org.apache.spark.util.{SerializableConfiguration, ShutdownHookManager, Utils}
+
+import scala.reflect.ClassTag
```
Could you update the import rule of you Intellj as
```
java

scala

othes

spark
```

davies commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/UnsafeRowParquetRecordReader.java:
```diff
@@ -0,0 +1,593 @@
+     */
+    if (requestedSchema.getFieldCount() == 0) {
+      // TODO: what does this mean?
+      throw new IOException("Empty request schema not supported.");
```
Should this be as easy as returning a empty UnsafeRow?

davies commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/UnsafeRowParquetRecordReader.java:
```diff
@@ -0,0 +1,593 @@
+    for (int i = 0; i < requestedSchema.getFieldCount(); ++i) {
+      Type t = requestedSchema.getFields().get(i);
+      if (!t.isPrimitive() || t.isRepetition(Type.Repetition.REPEATED)) {
+        throw new IOException("Complex types not supported.");
```
Should we use other Exception instead of IOException ?

davies commented on sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/UnsafeRowParquetRecordReader.java:
```diff
@@ -0,0 +1,613 @@
+      throw new IOException("expecting more rows but reached last block. Read "
+          + rowsReturned + " out of " + totalRowCount);
+    }
+    List<ColumnDescriptor> columns = requestedSchema.getColumns();
```
Should this be reused across groups?

davies replied:
nvm, the ColumnReader is related to pages

SparkQA commented:
**[Test build #46244 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46244/consoleFull)** for PR 9774 at commit [`1afe5dd`](https://github.com/apache/spark/commit/1afe5dd031ce15647df94dde71d1e39f7a8f2759).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `public abstract class SpecificParquetRecordReaderBase<T> extends RecordReader<Void, T> `\n  * `  abstract static class IntIterator `\n  * `  protected static final class ValuesReaderIntIterator extends IntIterator `\n  * `  protected static final class RLEIntIterator extends IntIterator `\n  * `  protected static final class NullIntIterator extends IntIterator `\n  * `public class UnsafeRowParquetRecordReader extends SpecificParquetRecordReaderBase<UnsafeRow> `\n

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46244/
Test PASSed.

rxin commented:
I'm merging this. Please address feedback in followup pull requests. Thanks!
